src/test/java/org/opensearch/security/transport/SecurityInterceptorTests.java

```diff
@@ -236,47 +237,54 @@ final void verifyOriginalContext(User user) {
         assertEquals(threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER), null);
     }
 
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    final CompletableFuture<TransportResponse> completableRequestDecorate(
+        AsyncSender sender,
+        Connection connection,
+        String action,
+        TransportRequest request,
+        TransportRequestOptions options,
+        TransportResponseHandler handler,
+        DiscoveryNode localNode
+    ) {
+        CompletableFuture<TransportResponse> future = new CompletableFuture<>();
+        securityInterceptor.sendRequestDecorate(sender, connection, action, request, options, handler, localNode);
+        future.thenRun(() -> { verifyOriginalContext(user); });
+        future.complete(null);
+        return future;
```

Comment:
Pretty sure this code doesn't work, the future needs to be completed by the asynchronous action in the 'sender', here is code that should handle the scenario for the happy path and report any exceptions raised that are catchable by the caller.  

```suggestion
        final var sendRequestFuture = new CompletableFuture<>();
        final var sender = new AsyncSender() {
            @Override
            public <T extends TransportResponse> void sendRequest(
                Connection connection,
                String action,
                TransportRequest request,
                TransportRequestOptions options,
                TransportResponseHandler<T> handler
            ) {
                // This is a Callable passed to this function - or any type that can execute and throw exceptions
                try {
                   verificationAction.call();
                   sendRequestFuture.complete(null);
                } catch (final Exception e) {
                   sendRequestFuture.completeExceptionally(e);
                }
            }
        };
        securityInterceptor.sendRequestDecorate(sender, connection, action, request, options, handler, localNode);
        sendRequestFuture.join()
```

Reply:
I am not sure I follow. ThenRun should not call verifyOriginalContext until the the future is completed. This is the documentation for thenRun: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#thenRun-java.lang.Runnable-

Why do you think this will not work?

Reply:
I am going to preemptively swap over to the form you suggested but still curious^

Reply:
>> Can you tell me if this code will / wont work as written?
>
> This Java code snippet involves using `CompletableFuture` and an asynchronous request. Let's analyze it for potential issues:
> 
> 1. **Creation of `CompletableFuture`**: The `CompletableFuture` is correctly instantiated. This part should work as expected.
> 
> 2. **Asynchronous Request Call**: The `securityInterceptor.sendRequestDecorate` method is called. This seems to be an asynchronous operation. However, it's not clear from this snippet whether this method is supposed to complete the `future` or not. If it's supposed to complete the `future`, it needs a reference to the `future` object to complete it, which is not passed in the snippet.
> 
> 3. **Chaining with `thenRun`**: The `thenRun` method is used to add an action (`verifyOriginalContext(user)`) that is triggered upon the future's completion. This is syntactically correct. However, it's important to note that `thenRun` runs in the same thread that completes the future, unless you explicitly use `thenRunAsync`.
> 
> 4. **Completing the Future**: The `future.complete(null)` immediately completes the future with a `null` value. This is a potential issue because it's called right after the asynchronous operation is initiated. If the intention is for `verifyOriginalContext(user)` to run after `sendRequestDecorate` completes its process, this immediate completion contradicts that. It effectively renders the asynchronous operation irrelevant, as the future is completed before the operation can potentially affect its outcome.
> 
> 5. **Return Statement**: The method correctly returns the `CompletableFuture`.
> 
> From this analysis, the main issue is with the premature completion of the future (`future.complete(null)`). This call should ideally be inside the callback of the asynchronous operation (`sendRequestDecorate`), assuming that method should dictate when the future is completed. If `sendRequestDecorate` does not interact with the `future`, then `future.complete(null)` might be correct, but in that case, the use of `CompletableFuture` might not be serving its intended purpose of managing asynchronous operations. 
> 
> In summary, the code will work syntactically, but it may not work as intended due to the premature completion of the `CompletableFuture`. The behavior depends on the internal workings of `sendRequestDecorate` and how it's supposed to interact with the future.

Sorry - stuck between meetings, I asked ChatGPT and I agree with its analysis